File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.empty()) return nullptr;
        return merge(lists, 0, lists.size() - 1);
    }

    ListNode* merge(vector<ListNode*>&lists, int l, int r) {
        if(l == r) {
            return lists[l];
        }
        if(l+1 == r) {
            return merge2Lists(lists[l], lists[r]);
        }

        int m = l + (r - l)/2;
        ListNode* left = merge(lists, l, m);
        ListNode* right = merge(lists, m + 1, r);

        return merge2Lists(left, right);
    }

    ListNode* merge2Lists(ListNode* a, ListNode* b) {
        ListNode dummy(0);
        ListNode* tail = &dummy;

        while(a && b) {
            if(a->val < b->val) {
                tail->next = a;
                a = a->next;
            } else {
                tail->next = b;
                b = b->next;
            }
            tail = tail->next;
        }

        tail->next = a ? a : b;
        return dummy.next;
    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        tail = dummy
        s, c = 0, 0

        a, b = l1, l2
        while a and b:
            s = a.val + b.val + c % 10
            d = s % 10
            tail.next = ListNode(d)
            c = s // 10
            a = a.next
            b = b.next
            tail = tail.next
        
        while a:
            s = c + a.val
            d = s % 10
            tail.next = ListNode(d)
            c = s // 10
            a = a.next
            tail = tail.next

        while b:
            s = c + b.val
            d = s % 10
            tail.next = ListNode(d)
            c = s // 10
            b = b.next
            tail = tail.next

        print(c)

        if c == 1:
            tail.next = ListNode(c)

        return dummy.next